### ISSUE TITLE
Broken root option behaviour after updating to Webpack 2

### ISSUE THREAD

**Issue by chrisnicola** [webpack/html-loader#91]
**I'm submitting a bug report**

**Webpack version:**
2.1.0-beta.22

**HTML-Loader version:**
0.4.4

**Please tell us about your environment:**
OSX 10.x

**Current behavior:**

Before updating to Webpack 2 setting the root option on my loader like so:

```
      { test: /\.slim$/, loader: 'ngtemplate!html?root=images!slm' },
```

Would cause something like this:

```
<img src="/image.png">
```

To resolve to `images/image.png` in the `resolve.root` folder. After it seems unable to resolve any of these root folder images. The error message is of the form below:

```
ERROR in ./webpack/templates/template.html.slim
Module not found: Error: Can't resolve '/images/images.png' in '~/project/webpack/templates'
```

**Expected/desired behavior:**

Ideally the same as before, it resolves the images to the root folder when there is a leading forward slash.

* **What is the motivation / use case for changing the behavior?**

Backwards compatibility
 
* **Language:** [TypeScript 2.0.2 ]

**Comment by chrisnicola** [comment]
This ended up being due to changes in the webpack config API. Hopefully before it is out of beta there will be warning letting people know about API changes from 1.0

**Comment by dsebastien** [comment]
Hello,

Just for my understanding: in Webpack 1.x, I didn't need any specific configuration for the html loader. Now that I use Webpack 2, I get errors like the one described here for templates where I use a root-relative path ("/...").

Do I need to add something in the configuration for html loader with Webpack 2? If so, what precisely? The "root" option?

I've just tried adding the root option but I just get different errors (parsing errors):
```
ERROR in ./~/stark/dist/src/ui/components/toast-notification/toast-notification.
component.html
Module build failed: Error: Parse Error: <div class=\"md-toast-content\"> <md-ic
on md-svg-icon=\"alert-circle\" ng-if=\"$ctrl.message.type === 0\"></md-icon> <m
d-icon md-svg-icon=\"alert\" ng-if=\"$ctrl.message.type === 1\"></md-icon> <md-i
con md-svg-icon=\"information\" ng-if=\"$ctrl.message.type === 2\"></md-icon> <s
pan class=\"md-toast-text\" role=\"alert\" aria-relevant=\"all\" aria-atomic=\"t
rue\" translate> {{ $ctrl.message.key }} </span> <md-button class=\"md-action\"
ng-if=\"$ctrl.message.actionLabel\" ng-click=\"$ctrl.closeToast()\" ng-class=\"$
ctrl.actionClasses\" translate> <span translate>{{ $ctrl.message.actionLabel }}<
/span> </md-button> </div> </md-toast> ";
    at new HTMLParser (C:\LocalData\DEV\Workspace\AAH-Stark\node_modules\html-mi
nifier\src\htmlparser.js:236:13)
    at minify (C:\LocalData\DEV\Workspace\AAH-Stark\node_modules\html-minifier\s
rc\htmlminifier.js:954:3)
    at Object.exports.minify (C:\LocalData\DEV\Workspace\AAH-Stark\node_modules\
html-minifier\src\htmlminifier.js:1307:10)
    at Object.module.exports (C:\LocalData\DEV\Workspace\AAH-Stark\node_modules\
html-loader\index.js:119:26)
 @ ./~/stark/dist/src/ui/components/toast-notification/toast-notification.compon
ent.js 3:19-65
 @ ./~/stark/dist/src/ui/services/toast-notification/toast-notification.service.
js
 @ ./~/stark/dist/src/ui/services/toast-notification/index.js
 @ ./~/stark/dist/src/ui/services/index.js
 @ ./~/stark/dist/src/ui/index.js
 @ ./~/stark/dist/src/index.js
 @ ./~/stark/dist/stark.js
 @ ./src/main.ts
 @ multi main
```

To be clear, I'm using require() to load my template files through the html loader and have them bundled with the rest of the code.

**Comment by michael-ciniawsky** [comment]
@dsebastien Can you show me a piece of the source HTML and the respective html-loader config in your current `webpack.config.js` ? For e.g does it work without HTMLMinifier enabled?

Maybe this is also helpful:
[Resolve#alias](https://webpack.js.org/configuration/resolve/#resolve-alias)

**Comment by dsebastien** [comment]
Here's the template it tries to load in this case:

````
<md-toast>
	<div class="md-toast-content">

		<md-icon md-svg-icon="alert-circle" ng-if="$ctrl.message.type === 0"></md-icon>
		<md-icon md-svg-icon="alert" ng-if="$ctrl.message.type === 1"></md-icon>
		<md-icon md-svg-icon="information" ng-if="$ctrl.message.type === 2"></md-icon>

		<span class="md-toast-text" role="alert" aria-relevant="all" aria-atomic="true" translate>
			{{ $ctrl.message.key }}
	    </span>
		
		<md-button class="md-action" ng-if="$ctrl.message.actionLabel" ng-click="$ctrl.closeToast()"
				   ng-class="$ctrl.actionClasses" translate>
			<span translate>{{ $ctrl.message.actionLabel }}</span>
		</md-button>
	</div>
</md-toast>
```

I just noticed though that I was using my prod configuration, so this one:
```
prod: {
				root: helpers.root(""),
				minimize: true,
				removeAttributeQuotes: false,
				caseSensitive: true,
				customAttrSurround: [
					[ /#/, /(?:)/ ],
					[ /\*/, /(?:)/ ],
					[ /\[?\(?/, /(?:)/ ]
				],
				customAttrAssign: [ /\)?\]?=/ ],
			},
```

**Comment by dsebastien** [comment]
Now I've fixed my configuration and the issue is a bit different.

I'm using the following config:
````
htmlLoader: {
			dev: {
				root: helpers.root(""),
			},
		
			prod: {
				root: helpers.root(""),
				minimize: true,
				removeAttributeQuotes: false,
				caseSensitive: true,
				customAttrSurround: [
					[ /#/, /(?:)/ ],
					[ /\*/, /(?:)/ ],
					[ /\[?\(?/, /(?:)/ ]
				],
				customAttrAssign: [ /\)?\]?=/ ],
			},
	},
```

Used like this with the LoaderOptionsPlugin of Webpack 2:
```
new LoaderOptionsPlugin({
			options: {
                                ...
				htmlLoader: commonData.htmlLoader.dev,
                                ...
```

And the loader is configured like this:
```
		{
			test: /\.html$/,
			use: [
				"html-loader",
			],
			exclude: [
				helpers.root("src/index.html"),
			],
		},
```

And now I get the following error:
```
ERROR in ./src/app/modules/request/components/comments-list/comments-list.compon
ent.html
Module not found: Error: Can't resolve './\""' in 'C:\LocalData\DEV\Workspace\AA
H-Stark\src\app\modules\request\components\comments-list'
 @ ./src/app/modules/request/components/comments-list/comments-list.component.ht
ml 1:411-430
 @ ./src/app/modules/request/components/comments-list/comments-list.component.ts

 @ ./src/app/modules/request/components/comments-list/index.ts
 @ ./src/app/modules/request/components/index.ts
 @ ./src/app/modules/request/module.ts
 @ ./src/app/modules/index.ts
 @ ./src/app/app.ts
 @ ./src/main.ts
 @ multi main
```

With this template:
```
<md-card class="comments-list">
	<md-card-title>
		<md-card-title-text>Comments</md-card-title-text>
	</md-card-title>
	<md-card-content>
		<md-list>
			<md-divider ></md-divider>
			<md-list-item class="comment-item md-3-line" ng-repeat="comment in $ctrl.commentHistory track by $index">
				<img src="/assets/images/avatar-default.png" class="md-avatar" alt="{{comment.user}}">
				<div class="md-list-item-text" layout="column">
					<h3>{{comment.user}}</h3>
					<p>{{comment.comment}}</p>
				</div>
				<div class="comment-date">
					2 days ago&nbsp;- {{comment.timestamp|date:"dd.MM.yyyy HH:mm"}}
				</div>
				<md-divider ></md-divider>
			</md-list-item>
		</md-list>
	</md-card-content>
</md-card>

```

So my root path seems to be used to adapt the img src attribute, but seems a bit broken.

I just don't understand well why/how the behaviour of the html-loader has changed between Webpack 1 & 2.

**Comment by dsebastien** [comment]
Stupid question for the errors related to img src: is it not possible to tell html-loader to just ignore the img src attribute?

**Comment by michael-ciniawsky** [comment]
```js
root: helpers.root("" // <---)
```

```bash
Module not found: Error: Can't resolve './\"" // <---'
```

```html
<img src="/assets/images/avatar-default.png" class="md-avatar" alt="{{comment.user}}">
```

Where is helpers.root defined ? Can you show me where it point to please ? If your assets are served later like e.g ```www.mysite.com/assets/images/*.png```  ---> output: { publicPath: 'www.mysite.com'} + htmlLoader: { root: '.' } ---> ```<img src="/assets/images/*.png>```

**Comment by dsebastien** [comment]
Hey @michael-ciniawsky 

Here's our helper function root: 
```
function root(args) {
    args = Array.prototype.slice.call(arguments, 0);
    return path.join.apply(path, [_root].concat(args));
}
```

where _root is `const _root = path.resolve(__dirname, ".."); // project root folder`

The complete helper file is here: https://github.com/dsebastien/angularjs-webpack-starter/blob/master/config/helpers.js

**Comment by michael-ciniawsky** [comment]
Why is `'..'` in `_root`? 'Normally' you got downwards form `__dirname ` not upwards '..', so that `__dirname === './'`
```js
const _root = path.resolve(___dirname) 
--> // C:\LocalData\DEV\Workspace\AAH-Stark\src\app ?

const _root = path.resolve(___dirname, '..') 
--> // C:\LocalData\DEV\Workspace\AAH-Stark\src ?
```

**Comment by michael-ciniawsky** [comment]
```js
const { resolve, join } = require('path')

const _root = (paths) => resolve(__dirname, Array.isArray(paths) ? join(...paths) : paths)
```

**Comment by michael-ciniawsky** [comment]
Why is ngTemplate-loader [using](https://github.com/dsebastien/angularjs-webpack-starter/blob/master/config/webpack.common.js#L162) html-loader at all :) ? According to it's [README](https://github.com/TheLarkInn/angular2-template-loader#example-markup) the usage of raw-loader is advised for both HTML && CSS alike.

**Comment by dsebastien** [comment]
Hey @michael-ciniawsky,

Sorry I left this aside for a while, now I'm back to it..

The code I gave was outdated; internally we have forked a long time ago. We don't even use ngTemplate-loader anymore :)

For the helpers, our code now looks like this:
```
const _root = path.resolve(); // project root folder
const _rootStark = path.resolve("node_modules/stark"); // stark root folder
```

Which is working perfectly for what we try to do.

Our goal is not to try and include the assets within the bundle, so if and when we use src=... in the html templates, we'd just like that to end up in the dom and have the client's browser go and fetch that resource, which will be available then.

We have defined the publicPath property.
